crates/subspace-networking/src/utils/unique_record_binary_heap.rs

```diff
@@ -1,65 +1,103 @@
 #[cfg(test)]
 mod tests;
 
+use crate::utils::multihash::ToMultihash;
 pub use libp2p::kad::record::Key;
 use libp2p::kad::KBucketDistance;
 pub use libp2p::PeerId;
 use std::cmp::Ordering;
 use std::collections::BTreeSet;
+use subspace_core_primitives::PieceIndex;
 
 type KademliaBucketKey<T> = libp2p::kad::KBucketKey<T>;
 
 // Helper structure. It wraps Kademlia distance to a given peer for heap-metrics.
 #[derive(Debug, Clone)]
-struct RecordHeapKey {
+struct RecordHeapKey<K> {
     peer_distance: KBucketDistance,
-    key: KademliaBucketKey<Key>,
+    key: K,
 }
 
-impl RecordHeapKey {
+impl<K> RecordHeapKey<K>
+where
+    Key: From<K>,
+    K: Clone,
+{
     fn peer_distance(&self) -> KBucketDistance {
         self.peer_distance
     }
 
-    fn new(peer_key: &KademliaBucketKey<PeerId>, key: KademliaBucketKey<Key>) -> Self {
-        let peer_distance = key.distance(peer_key);
+    fn new(peer_key: &KademliaBucketKey<PeerId>, key: K) -> Self {
+        let peer_distance = KademliaBucketKey::new(Key::from(key.clone())).distance(peer_key);
         Self { peer_distance, key }
     }
 }
 
-impl Eq for RecordHeapKey {}
+impl<K> Eq for RecordHeapKey<K>
+where
+    Key: From<K>,
+    K: Clone,
+{
+}
 
-impl PartialEq<Self> for RecordHeapKey {
+impl<K> PartialEq<Self> for RecordHeapKey<K>
+where
+    Key: From<K>,
+    K: Clone,
+{
     fn eq(&self, other: &Self) -> bool {
         self.peer_distance().eq(&other.peer_distance())
     }
 }
 
-impl PartialOrd<Self> for RecordHeapKey {
+impl<K> PartialOrd<Self> for RecordHeapKey<K>
+where
+    Key: From<K>,
+    K: Clone,
+{
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
         Some(self.cmp(other))
     }
 }
 
-impl Ord for RecordHeapKey {
+impl<K> Ord for RecordHeapKey<K>
+where
+    Key: From<K>,
+    K: Clone,
+{
     fn cmp(&self, other: &Self) -> Ordering {
         self.peer_distance().cmp(&other.peer_distance())
     }
 }
 
+/// Wrapper data structure that allows to work with keys as Kademlia keys, while not storing 32
+/// bytes if the key itself is smaller, while potentially trading a bit of runtime performance.
+#[derive(Debug, Copy, Clone)]
+pub struct UniqueRecordBinaryHeapWrapper<T>(pub T);
```

Comment:
It doesn't seem to be a heap wrapper judging by the comment and conversion trait. Do you mind picking a different name? Variants: PieceIndexHeapWrapper or PieceIndexKeyWrapper.

Reply:
It has nothing to do with piece index necessarily. Yes, it has convenient implementation for piece index, but it works with any type and because of that is generic.

Reply:
`KeyTypeWrapper` ? The current name suggests that we wrap the heap.

Reply:
Makes sense, went with `KeyWrapper`